R/type2.R

```diff
@@ -125,55 +125,34 @@ vec_default_ptype2 <- function(x,
 
   opts <- match_fallback_opts(...)
 
-  # If both data frames, first find common type of columns before the
-  # same-type fallback
-  if (df_needs_normalisation(x, y, opts)) {
-    out <- vec_ptype2_df_fallback_normalise(
+  if (opts$s3_fallback && can_fall_back_2(x, y)) {
+    common <- common_class_suffix(x, y)
+    if (length(common)) {
+      return(new_common_class_fallback(x, common))
+    }
+  }
+
+  if (is.data.frame(x) && is.data.frame(y)) {
+    out <- vec_ptype2_df_fallback(
       x,
       y,
       opts,
       x_arg = x_arg,
       y_arg = y_arg,
       call = call
     )
-    x <- out$x
-    y <- out$y
-  }
 
-  if (opts$s3_fallback && can_fall_back_2(x, y)) {
-    common <- common_class_suffix(x, y)
-    if (length(common)) {
-      return(new_common_class_fallback(x, common))
+    if (identical(non_df_attrib(x), non_df_attrib(y))) {
+      attributes(out) <- c(df_attrib(out), non_df_attrib(x))
     }
+
```

Comment:
The reason I'm not seeing a bare-df fallback case here (unlike the cast path in https://github.com/r-lib/vctrs/pull/1733/files#diff-d24441095488b3e181d91060e43aa207b655e9070928ffc20ab5d9351d3ef2c7R212) is because it is already handled in `vec_ptype2_df_fallback()` right?

Reply:
This creates a bare df common type (bare-type fallback) and restores attributes if possible (same-type callback).